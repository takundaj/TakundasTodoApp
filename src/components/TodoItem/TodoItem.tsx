import { Animated, Text, TouchableOpacity, Alert } from "react-native";
import React, { useRef } from "react";
import Ionicons from "@expo/vector-icons/Ionicons";
import { Todo } from "../../types/data";
import { Swipeable } from "react-native-gesture-handler";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { StackParamList } from "../../types/navigationTypes";
import { useNavigation } from "@react-navigation/native";
import { FirebaseFirestoreTypes } from "@react-native-firebase/firestore";
import styles from "./todoItem.styles";

type Props = {
  title?: string;
  isChecked: boolean;
  onCompleteTodo: (newTodo: Todo) => void;
  todos?: FirebaseFirestoreTypes.DocumentData[];
  deleteTodo: () => void;
  color: string;
  groupId?: string;
};

type TodoItemNavigationProp = NativeStackNavigationProp<
  StackParamList,
  "TodoModal"
>;

const TodoItem = ({
  title,
  groupId,
  isChecked,
  onCompleteTodo,
  todos,
  deleteTodo,
  color,
}: Props) => {
  const navigation = useNavigation<TodoItemNavigationProp>();

  const swipeableRef = useRef<Swipeable>(null);

  const triggerAlert = () => {
    Alert.alert("Are you sure you want to delete this item?", "", [
      {
        text: "Cancel",
        onPress: () => {},
        style: "cancel",
      },
      {
        text: "OK",
        onPress: () => deleteTodo(),
      },
    ]);

    swipeableRef.current?.close();
  };

  return (
    <Swipeable
      ref={swipeableRef}
      renderRightActions={() => (
        <TouchableOpacity
          style={styles.swipeActionRight}
          onPress={triggerAlert}
        >
          <Ionicons name="trash" size={24} color={"red"} />
        </TouchableOpacity>
      )}
      renderLeftActions={() => (
        <TouchableOpacity
          style={styles.swipeActionLeft}
          onPress={() => {
            navigation.navigate("TodoModal", { title, groupId, todos });
            swipeableRef?.current?.close();
          }}
        >
          <Ionicons name="md-pencil" size={24} color={"orange"} />
        </TouchableOpacity>
      )}
    >
      <Animated.View style={styles.swipeableContainer}>
        <Text>{title}</Text>
        <TouchableOpacity
          onPress={() =>
            title ? onCompleteTodo({ title: title, isDone: !isChecked }) : null
          }
        >
          <Ionicons
            name={isChecked ? "checkmark-circle" : "checkmark-circle"}
            size={30}
            color={isChecked ? color || "turquoise" : "lightgrey"}
          />
        </TouchableOpacity>
      </Animated.View>
    </Swipeable>
  );
};

export default TodoItem;
