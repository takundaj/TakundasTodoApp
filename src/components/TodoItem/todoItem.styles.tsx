import { StyleSheet } from "react-native";
import { size } from "../../utils/helpers";

export default StyleSheet.create({
  swipeActionRight: {
    justifyContent: "center",
    alignItems: "center",
    paddingRight: size(2),
  },
  swipeActionLeft: {
    justifyContent: "center",
    alignItems: "center",
    paddingLeft: size(4),
  },
  swipeableContainer: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    padding: size(4),
    marginVertical: size(2),
    borderRadius: size(2),
    backgroundColor: "white",
    margin: size(4),
    shadowOffset: {
      width: 0.5,
      height: 1,
    },
    shadowOpacity: 0.8,
    shadowRadius: 2,
  },
});
