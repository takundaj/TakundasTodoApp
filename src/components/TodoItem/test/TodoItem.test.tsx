import { render, fireEvent } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import TodoItem from "../TodoItem";

const mockProps = {
  title: "dummyTitle",
  groupId: "dummyId",
  isChecked: false,
  onCompleteTodo: jest.fn(),
  todos: [{ title: "dummyTitle", createdAt: "", isDone: false }],
  deleteTodo: jest.fn(),
  color: "red",
};

jest.mock("@react-navigation/core", () => ({
  ...jest.requireActual("@react-navigation/core"),
  useNavigation: jest.fn(),
}));

test("Should render GroupCard component", () => {
  const { getByTestId } = render(<TodoItem {...mockProps} />);

  const container = getByTestId("todo-item-container");
  expect(container).toBeDefined();
});

test("Should call onCompleteTodo after icon press", () => {
  const { getByTestId } = render(<TodoItem {...mockProps} />);

  const icon = getByTestId("todo-item-icon");
  fireEvent.press(icon);
  expect(mockProps.onCompleteTodo).toHaveBeenCalled();
});

test("Should match snapshot", () => {
  const tree = renderer.create(<TodoItem {...mockProps} />);
  expect(tree).toMatchSnapshot();
});
