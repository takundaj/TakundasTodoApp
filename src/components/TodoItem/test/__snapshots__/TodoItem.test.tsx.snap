// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<View
  collapsable={false}
  handlerTag={5}
  handlerType="PanGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  style={
    {
      "overflow": "hidden",
    }
  }
  testID="todo-item-container"
>
  <View
    collapsable={false}
    style={
      {
        "bottom": 0,
        "flexDirection": "row",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": [
          {
            "translateX": -10000,
          },
        ],
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "opacity": 1,
          "paddingLeft": 16,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "orange",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
  <View
    collapsable={false}
    style={
      {
        "bottom": 0,
        "flexDirection": "row-reverse",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": [
          {
            "translateX": -10000,
          },
        ],
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "opacity": 1,
          "paddingRight": 8,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "red",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
  <View
    collapsable={false}
    handlerTag={6}
    handlerType="TapGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    pointerEvents="auto"
    style={
      {
        "transform": [
          {
            "translateX": -0,
          },
        ],
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 8,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 16,
          "marginVertical": 8,
          "padding": 16,
          "shadowOffset": {
            "height": 1,
            "width": 0.5,
          },
          "shadowOpacity": 0.8,
          "shadowRadius": 2,
        }
      }
    >
      <Text>
        dummyTitle
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="todo-item-icon"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "lightgrey",
                "fontSize": 30,
              },
              undefined,
              {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
